### PR TITLE
[5.1] Fix postgres union clause

### PR DESCRIPTION
**Pull request by ghost**
Re-request for 5.1.
See https://github.com/laravel/framework/pull/12277

### CODE REVIEW

File: src/Illuminate/Database/Query/Grammars/PostgresGrammar.php [hunk 1]
```diff
@@ -242,4 +255,23 @@ public function compileTruncate(Builder $query)
     {
         return ['truncate '.$this->wrapTable($query->from).' restart identity' => []];
     }
+
+    /**
+     * Concatenate an array of segments, removing empties.
+     *
+     * @param  array   $segments
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Query/Grammars/PostgresGrammar.php]: cs

File: src/Illuminate/Database/Query/Grammars/PostgresGrammar.php [hunk 2]
```diff
@@ -242,4 +255,23 @@ public function compileTruncate(Builder $query)
...
+    protected function concatenate($segments)
+    {
+        if (! empty($segments['unions']) && (! empty($segments['orders']) || ! empty($segments['limit']))) {
+            array_unshift($segments, '(');
+            $unionIndex = array_search('unions', array_keys($segments), true);
+            array_splice($segments, $unionIndex, 0, ')');
+        }
+
+        return implode(' ', array_filter($segments, function ($value) {
+            return (string) $value !== '';
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Query/Grammars/PostgresGrammar.php]: Isn't the casting evaluated on the whole thing, not `$value`?

Comment (segfall) [inline src/Illuminate/Database/Query/Grammars/PostgresGrammar.php]: cast applies to $value only. Return block copied from the base class unchanged. Made a change to use parent impl.

Only new code is the `if {..}` block, which adds ( as first element and ) before union, wrapping all expression.

File: src/Illuminate/Database/Query/Grammars/PostgresGrammar.php [hunk 3]
```diff
@@ -242,4 +255,21 @@ public function compileTruncate(Builder $query)
...
         return ['truncate '.$this->wrapTable($query->from).' restart identity' => []];
     }
+
+    /**
+     * Concatenate an array of segments, removing empties.
+     *
+     * @param  array   $segments
+     * @return string
+     */
+    protected function concatenate($segments)
```
Comment (taylorotwell) [inline src/Illuminate/Database/Query/Grammars/PostgresGrammar.php]: This entire function needs some explanation. It's hard to follow and is unclear what is going on.

**Comment by segfall** [comment]
Uh. Well, there's a bug, and this is an easiest fix I came up with. I agree it may not be particularly readable, and if there's a better solution, please implement it instead of this one.

This function turns an array of query components (['select' => '...', 'where' => '...']) extracted from Query\Builder into an sql string.

When using UNION operator with sub-statements containing LIMIT or OrderBy clauses, they must be wrapped in parens.

Second and following sub-statements are parenthesized in the compileUnion() method,see MySqlGrammar class. But this method does not deal with the very first sub-statement, which is everything from SELECT to first UNION keyword, which must be wrapped, too.

So when concatenating, left paren is shifted onto the array, and right paren is inserted just before UNION array key. The condition tests if parens are needed.

**Comment by taylorotwell** [comment]
I wrote a test for your change and get the following failure:

http://d.pr/i/1fa0K/2bKSmSGu

Is the first select statement not supposed to be wrapped for Postgres?

**Comment by segfall** [comment]
A statement _needs_ to be wrapped only in the presence of LIMIT or ORDER BY clauses, so I opted to wrap it only when required. It _may_ be wrapped in all cases without errors.

Same rules apply for the statements that follow, but I copied MySQL implementation.
